File: backend/app/schemas/plugin_repo.py
"""插件仓库 Pydantic schemas。"""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class PluginRepoCredentialUpdate(BaseModel):
    """更新仓库凭证请求体。空 token 表示清除凭证。"""

    auth_type: str | None = Field(default="github_token", description="凭证类型；当前支持 github_token 或 none")
    token: str | None = Field(default=None, description="GitHub fine-grained / classic token；不会回显")

    @field_validator("auth_type")
    @classmethod
    def _normalize_auth_type(cls, value: str | None) -> str:
        raw = str(value or "github_token").strip().lower()
        if raw in {"", "none", "public"}:
            return "none"
        if raw in {"github", "github_token", "token", "pat"}:
            return "github_token"
        raise ValueError("auth_type 仅支持 github_token 或 none")


class PluginRepoCreate(BaseModel):
    """新增仓库请求体。

    ``url`` 必填；``name``/``description`` 可选，缺省时服务端会从 URL 派生显示名。
    """

    url: str = Field(..., description="git URL，如 https://github.com/foo/bar.git")
    name: str | None = Field(default=None, description="展示名；为空时从 URL 派生")
    description: str | None = Field(default=None, description="可选备注")
    credential: PluginRepoCredentialUpdate | None = Field(
        default=None,
        description="可选私有 GitHub 仓库凭证；token 加密保存且不会回显",
    )


class PluginRepoOut(BaseModel):
    """仓库 DB 行的对外形态。"""

    id: int
    name: str
    url: str
    description: str
    auth_type: str = "none"
    has_credentials: bool = False
    added_at: datetime | None = None
    updated_at: datetime | None = None

    class Config:
        from_attributes = True


class PluginRepoPlugin(BaseModel):
    """仓库内单个插件的描述（仅来自 plugin.json 静态解析，无执行）。

    ``installed`` 由服务层根据 ``remote_plugin.name`` 是否已存在来填，便于
    前端给出“已安装/可安装”按钮状态。
    """

    name: str
    display_name: str = ""
    description: str = ""
    author: str = ""
    version: str = "0.0.0"
    installed: bool = False
    installed_version: str | None = None
    update_available: bool = False
    # 该插件在仓库内的相对子目录（用于安装时定位）；若插件位于仓库根目录则为 ""
    subdir: str = ""

File: frontend/src/types/pluginRepo.ts
export interface PluginRepo {
  id: number;
  name: string;
  url: string;
  description: string;
  auth_type: "none" | "github_token" | string;
  has_credentials: boolean;
  added_at: string | null;
  updated_at: string | null;
}

export interface PluginRepoCredentialUpdate {
  auth_type?: "none" | "github_token" | string | null;
  token?: string | null;
}

export interface PluginRepoCreate {
  url: string;
  name?: string | null;
  description?: string | null;
  credential?: PluginRepoCredentialUpdate | null;
}

export interface PluginRepoPlugin {
  name: string;
  display_name: string;
  description: string;
  author: string;
  version: string;
  installed: boolean;
  installed_version?: string | null;
  update_available?: boolean;
  subdir: string;
}

export interface InstallFromRepoBody {
  default_enabled?: boolean;
}
